auth/postgres/repo.go

```diff
@@ -0,0 +1,507 @@
+		ID:   patID,
+	}
+
+	_, err := pr.db.NamedQueryContext(ctx, q, upm)
+	if err != nil {
+		return errors.Wrap(repoerr.ErrUpdateEntity, err)
+	}
+
+	return nil
+}
+
+func (pr *patRepo) Remove(ctx context.Context, userID, patID string) error {
+	q := `DELETE FROM pats WHERE user_id = :user_id AND id = :id`
+	upm := dbPagemeta{
+		User: userID,
+		ID:   patID,
+	}
+
+	_, err := pr.db.NamedQueryContext(ctx, q, upm)
+	if err != nil {
+		return postgres.HandleError(repoerr.ErrRemoveEntity, err)
+	}
+
+	return nil
+}
+
+func (pr *patRepo) AddScopeEntry(ctx context.Context, userID string, scopes []auth.Scope) error {
+	dbscopes := toDBScope(scopes)
+	_, err := pr.db.NamedQueryContext(ctx, saveScopeQuery, dbscopes)
+	if err != nil {
+		return postgres.HandleError(repoerr.ErrUpdateEntity, err)
+	}
+
+	if err := pr.cache.Save(ctx, scopes); err != nil {
+		return errors.Wrap(repoerr.ErrUpdateEntity, err)
+	}
+
+	return nil
+}
+
+func (pr *patRepo) RemoveScopeEntry(ctx context.Context, userID string, scopes []auth.Scope) error {
+	dbscopes := toDBScope(scopes)
+
+	_, err := pr.db.NamedQueryContext(ctx, deleteScopesQuery, dbscopes)
+	if err != nil {
+		return errors.Wrap(repoerr.ErrUpdateEntity, err)
+	}
+
+	if err := pr.cache.Remove(ctx, scopes); err != nil {
+		return errors.Wrap(repoerr.ErrRemoveEntity, err)
+	}
+
+	return nil
+}
+
+func (pr *patRepo) CheckScopeEntry(ctx context.Context, userID, patID string, entityType auth.EntityType, optionalDomainID string, operation auth.Operation, entityID string) error {
+	authorized := pr.cache.CheckScope(ctx, patID, optionalDomainID, entityType, operation, entityID)
+	if authorized {
+		return nil
+	}
+
+	scope := dbScope{
+		PatID:            patID,
+		EntityType:       entityType.String(),
+		OptionalDomainID: optionalDomainID,
+		Operation:        operation.String(),
+		EntityID:         entityID,
+	}
+
+	rows, err := pr.db.NamedQueryContext(ctx, checkScopeQuery, scope)
+	if err != nil {
+		return errors.Wrap(repoerr.ErrViewEntity, err)
+	}
+	defer rows.Close()
+
+	if rows.Next() {
+		var sc dbScope
+		if err := rows.StructScan(&sc); err != nil {
+			return errors.Wrap(repoerr.ErrViewEntity, err)
+		}
+
+		entityType, err := auth.ParseEntityType(sc.EntityType)
+		if err != nil {
+			return errors.Wrap(repoerr.ErrViewEntity, err)
+		}
+		operation, err := auth.ParseOperation(sc.Operation)
+		if err != nil {
+			return errors.Wrap(repoerr.ErrViewEntity, err)
+		}
+		authScope := auth.Scope{
+			PatID:            sc.PatID,
+			OptionalDomainID: sc.OptionalDomainID,
+			EntityType:       entityType,
+			EntityID:         sc.EntityID,
+			Operation:        operation,
+		}
+
+		if err := pr.cache.Save(ctx, []auth.Scope{authScope}); err != nil {
+			log.Println(err)
+		}
+
+		if authScope.Authorized(entityType, optionalDomainID, operation, entityID) {
+			return nil
+		}
+	}
+
+	return repoerr.ErrNotFound
+}
+
+func (pr *patRepo) RemoveAllScopeEntry(ctx context.Context, patID string) error {
+	pm := dbPagemeta{
```

Comment:
If we add userID in cache key, then here we need `userID`
```suggestion
func (pr *patRepo) RemoveAllScopeEntry(ctx context.Context, userID, patID string) error {
```